### PR TITLE
Add add_constraint method to SingleTableMetadata

### LINKED ISSUE #881: Add add_constraint method to SingleTableMetadata
### Problem Description

_As a user, it would be very useful to be able to add constraints to the metadata programmatically since some of them often get missed the first time around._

### Expected behavior

- Add `add_constraint` method
    - This should add the constraint info to the dict
    - Parameters
        - `constraint_name`: string that is the name of the constraint class
        - `**kwargs`: Any other required key word arguments for the constraint.

```python
>>> metadata.add_constraint(
        constraint_name='FixedIncrements',
        column_name='salary',
        increment=5000)

# we should do a validation when they add this
>>> metadata.add_constraint(
        constraint_name='Inequality',
        low_column_name='child_age',
        high_column_name='start_date')
Error: An Inequality constraint is being applied to mismatched sdtypes ("child_age", "start_date"). Both columns must be either numerical or datetime.
```

### Additional context

- For validation and error handling use the `_validate_inputs` method made in #878 and in #879

### PR DESCRIPTION
**Pull request by amontanez24**
resolves #881

### CODE REVIEW

File: sdv/constraints/tabular.py [hunk 1]
```diff
@@ -1139,6 +1273,24 @@ def __init__(self, column_names):
...
+            primary_key = tuple(metadata._primary_key)
+
+        for column in column_names:
+            if column in metadata._alternate_keys or column in primary_key:
+                keys.append(column)
+
+        raise ConstraintMetadataError(
+            f"A Unique constraint is being applied to columns '{keys}'. "
+            'These columns are already a key for that table.'
+        )
```
Comment (pvk-developer) [inline sdv/constraints/tabular.py:1292]: Shouldn't this be raised in case that there are `keys` ?

File: sdv/constraints/base.py [hunk 2]
```diff
@@ -125,7 +126,59 @@ def _validate_inputs(cls, **kwargs):
...
+    @classmethod
+    def _validate_metadata_specific_to_constraint(cls, metadata, **kwargs):
+        pass
+
+    @classmethod
+    def _validate_metadata(cls, metadata, **kwargs):
+        """Validate the metadata against the constraint.
+
+        Args:
+            metadata (dict):
```
Comment (pvk-developer) [inline sdv/constraints/base.py:155]: `sdv.metadata.SingleTableMetadata` instead of dict.

File: sdv/constraints/base.py [hunk 3]
```diff
@@ -272,6 +325,16 @@ def filter_valid(self, table_data):
...
         return table_data[valid]
 
+    @classmethod
+    def _get_class_from_dict(cls, constraint_class):
+        subclasses = get_subclasses(cls)
+        if isinstance(constraint_class, str):
+            if '.' in constraint_class:
+                constraint_class = import_object(constraint_class)
+            else:
+                constraint_class = subclasses[constraint_class]
```
Comment (pvk-developer) [inline sdv/constraints/base.py:335]: break line

File: sdv/constraints/tabular.py [hunk 4]
```diff
@@ -339,6 +340,30 @@ def _validate_init_inputs(low_column_name, high_column_name, strict_boundaries):
...
 
+    @classmethod
+    def _validate_metadata_columns(cls, metadata, **kwargs):
+        column_names = [kwargs.get('high_column_name'), kwargs.get('low_column_name')]
+        missing_columns = [column for column in column_names if column not in metadata._columns]
+        if missing_columns:
+            raise ConstraintMetadataError(
+                f' A {cls.__name__} constraint is being applied to invalid column names '
+                f'{missing_columns}. The columns must exist in the table.'
+            )
```
Comment (pvk-developer) [inline sdv/constraints/tabular.py:351]: Could we do something like this:
```python
kwargs['column_names'] = [kwargs.get('high_column_name'), kwargs.get('low_column_name')]
super()._validate_metadata_columns(metadata, **kwargs)
```

File: sdv/constraints/tabular.py [hunk 5]
```diff
@@ -670,6 +739,35 @@ class Range(Constraint):
...
+        high = kwargs.get('high_column_name')
+        low = kwargs.get('low_column_name')
+        middle = kwargs.get('middle_column_name')
+        column_names = [high, low, middle]
+        missing_columns = [column for column in column_names if column not in metadata._columns]
+        if missing_columns:
+            raise ConstraintMetadataError(
+                f' A {cls.__name__} constraint is being applied to invalid column names '
+                f'{missing_columns}. The columns must exist in the table.'
+            )
```
Comment (pvk-developer) [inline sdv/constraints/tabular.py:753]: Could we do something like this:
```python
high = kwargs.get('high_column_name')
low = kwargs.get('low_column_name')
middle = kwargs.get('middle_column_name')
kwargs['column_names'] = [high, low, middle]
super()._validate_metadata_columns(metadata, **kwargs)
```

File: tests/unit/constraints/test_base.py [hunk 6]
```diff
@@ -158,6 +159,42 @@ def test_import_object_function():
 
 class TestConstraint():
 
+    @patch('sdv.constraints.base.Constraint._validate_inputs')
+    @patch('sdv.constraints.base.Constraint._validate_metadata_columns')
+    @patch('sdv.constraints.base.Constraint._validate_metadata_specific_to_constraint')
+    def test__validate_metadata(
+            self, validate_metadata_columns_mock, validate_metadata_specific_to_constraint_mock,
+            validate_inputs_mock):
+        """Test the ``_validate_metadata method.
```
Comment (pvk-developer) [inline tests/unit/constraints/test_base.py:168]: missing `` at the end of _validate_metadata

File: tests/unit/constraints/test_base.py [hunk 7]
```diff
@@ -158,6 +159,42 @@ def test_import_object_function():
...
+
+        The method should compile the error messages returned from ``_validate_inputs``,
+        ``_validate_metadata_columns and ``_validate_metadata_specific_to_constraint``
```
Comment (pvk-developer) [inline tests/unit/constraints/test_base.py:171]: missing `` at _validate_metadata_columns

**Review by pvk-developer — CHANGES_REQUESTED** [review]

File: sdv/constraints/tabular.py (hunk 1 above)
Comment (amontanez24) [reply sdv/constraints/tabular.py:1292]: Isn't that what this is doing?

Comment (amontanez24) [reply sdv/constraints/tabular.py:1292]: Yup, I forgot to test this one too

**Review by pvk-developer — APPROVED** [review]
Thanks for addressing the comments. :shipit:

**Review by fealho — APPROVED** [review]

**Pull request merged**